### ISSUE TITLE
Runnable LFRIc examples

### ISSUE THREAD

**Issue by hiker** [stfc/PSyclone#745]
I've suggested during our telco that I would add a runnable LRic example, i.e. using a 'reduced' version of the LFRic infrastructure (similar to eg5/6 of gocean, e.g. to show how profile and extraction works). This would be useful for people who like to 'play' with PSyclone in order to better understand it, and to develop PSyData-based tools for LFRic. I thought we had a ticket for that, but I couldn't find it now.

I have started with that, but am not sure about the best scope for this reduced infrastructure:

1. A minimal example (no MPI) that just allows us to create a mesh, vector spaces, then fields, and some kernel calls (based on the code I have studies so far, it probably needs a bit more ;) ) 
    - Advantage: small additional code base
    - Advantage: easy to understand (and accidentally a nice intro to LFRic).
    - Disadvantage: we need to edit the infrastructure library to remove unnecessary dependencies, which might mean additional (duplicated??) work if there are changes to the infrastructure (that PSyclone is using).
    - Disadvantage: We can't run examples using MPI (and related transformations)

2. A minimal example with MPI
    - Advantage: as 1.
    - Advantage: we can use MPI transformations
    - Disadvantage: as above
    - Disadvantage: even more files need to be added to handle domain decomposition.
    - Disadvantage: we need to implement our own MPI wrapper as mpi_mod (or else we depend on yaxt).

3. We add the full LFRic-only infrastructure, but without MPI
    - Advantage: Reasonable easy to keep it up-to-date
    - Disadvantage: difficult to understand (large code size)
    - Disadvantage: we still need to remove MPI, and some files do have a hard-coded dependency on the MPI module (e.g. field_mod uses global_sum from mpi_mod, which in turn directly depends on MPI and yaxt as far as I can see), which means potentially more work.

4. We add the full LFRic infrastructure and dependencies (at least yaxt, likely netcdf, ... and probably more)
    - Advantage: easy to keep up-to-date
    - Disadvantage: somewhat of a nightmare to build
    - Disadvantage: somewhat difficult build system 

I have to admit that my favourite is solution 1) - just ignore MPI, remove unnecessary code to reduce complexity - but I thought I'd better ask you all (@rupertford , @arporter, @sergisiso, @TeranIvy , @christophermaynard ) for feedback.

We could try to add #ifdef around infrastructure files that we use, and feed them back to LFRic, but I am not sure that this is something they would even want to do in the first place.

Suggestions welcome :)

**Comment by arporter** [comment]
Personally, I'm not keen on any of these solutions as I suspect that even a minimal infrastructure would require a fair amount of effort. However, I may be wrong about that - I guess it depends on how much is needed in order to actually run a kernel?

What about using the containerised LFRic infrastructure? That would solve a lot of these problems and avoids us sinking effort into the 'stub' infrastructure. Presumably that would require people to have an Intel compiler license but that seems a small limitation (and a gfortran container could be built?).

Finally, how 'mini' are the LFRic 'mini-apps'? Should we be using one of them? (Question for the LFRicians really.)

**Comment by hiker** [comment]
Re miniapp: I have just started to compile 'skeleton'. As part of the build process it copies over 400 f90 files, totaling over 100k lines into a working directory. I somehow doubt that they are all strictly required, but that's the files I see during a build.

If we are happy to ignore MPI, the way a truly minimal example would look like I _think_ is something like this (am still studying the details of the code):
```
  global_mesh = create_global_mesh(...)
  mesh = create_mesh(global_mesh, ...)
  vector_space = function_space_type(mesh, ...)
  field%initialise(vector_space, ...);

   ! do some nifty stuff with kernels and psyclone here :)
```
One annoying issue is that most function pass IDs around (which requires all kind of container classes), but I think this can be changed by passing pointers to the mesh, as shown above (I guess this is done in the big apps to reduce dependencies - an integer being passed around does not need to trigger a recompile when the user-defined type is changed). Now, the devil is in the details, of course, but here is what is already working:
```
    type(global_mesh_type) :: global_mesh
    global_mesh = global_mesh_type("ignore-filename","ignore-mesh-name")
```
Yes, only a tiny bit, and still wip, but a start :)

**Comment by arporter** [comment]
:-) re miniapp - possibly it needs renaming!? I'm just concerned that this is going to be a considerable draw on resources and we are not LFRic experts and thus not best placed to do it.

**Comment by TeranIvy** [comment]
Re separating infrastructure, LFRicians realised that we have general mathematics functionality which should also be separated into a library. In fact, this is what prevents us from getting rid of all calls to `gungho` in building miniapps. The ticket is [2175](https://code.metoffice.gov.uk/trac/lfric/ticket/2175).

**Comment by TeranIvy** [comment]
Merged #954 so closing this issue. Any further refactoring can be done in a separate issue, especially given the ongoing changes in LFRic infrastructure regarding how meshes are created.
